File: alphabet_symmetry.rb
# 7 kyu
# Alphabet symmetry

# Consider the word "abode". We can see that the letter a is in position 1 and b 
# is in position 2. In the alphabet, a and b are also in positions 1 and 2. 
# Notice also that d and e in abode occupy the positions they would occupy in 
# the alphabet, which are positions 4 and 5.

# Given an array of words, return an array of the number of letters that occupy 
# their positions in the alphabet for each word.
# Input will consist of alphabet characters, both uppercase and lowercase.No spaces.



# iterate thru the array, map, transfer each word into array
  # for each word subarray, using each_with_index to iterate
    # - if the char's index == LOWERCASE.index(char), count + 1

LOWERCASE = ('a'..'z').to_a

def symmetry_letters(array)
  result = []
  array.map(&:chars).each do |subarray|
  count = 0
    subarray.each_with_index do |char, index|
      count += 1  if index == LOWERCASE.index(char.downcase) 
    end
  result << count
  end
  p result
end

symmetry_letters(["abode","ABc","xyzD"]) == [4, 3, 1]

# --- smarter sol ---

def solve(strings)
  strings.map{ |s| s.downcase.each_char.zip('a'..'z').count{ |a, b| a == b } }
end

# zip is a VERY important and handy way to map letters to number
# for example:
# ('a'..'z').to_a.zip(1..26)
# [["a", 1], ["b", 2], ["c", 3], ["d", 4], ["e", 5], ["f", 6], ["g", 7], 
# ["h", 8], ["i", 9], ["j", 10], ["k", 11], ["l", 12], ["m", 13], ["n", 14], 
# ["o", 15], ["p", 16], ["q", 17], ["r", 18], ["s", 19], ["t", 20], ["u", 21], 
# ["v", 22], ["w", 23], ["x", 24], ["y", 25], ["z", 26]] 

# then transfer it to hash looks more straightforward
# ('a'..'z').to_a.zip(1..26).to_h
# {"a"=>1, "b"=>2, "c"=>3, "d"=>4, "e"=>5, "f"=>6, "g"=>7, "h"=>8, 
#   "i"=>9, "j"=>10, "k"=>11, "l"=>12, "m"=>13, "n"=>14, "o"=>15, 
#   "p"=>16, "q"=>17, "r"=>18, "s"=>19, "t"=>20, "u"=>21, "v"=>22, 
#   "w"=>23, "x"=>24, "y"=>25, "z"=>26} 

File: JavaScript/alphabet_symmetry.js
// 7 kyu
// Alphabet symmetry

// Consider the word "abode". We can see that the letter a is in position 1 and b 
// is in position 2. In the alphabet, a and b are also in positions 1 and 2. 
// Notice also that d and e in "abode" occupy the positions they would occupy in 
// the alphabet, which are positions 4 and 5.

// Given an array of words, return an array of the number of letters that occupy 
// their positions in the alphabet for each word.
// Input will consist of alphabet characters, both uppercase and lowercase. No spaces.

// input:
//   - an array of words
// output:
//   - an array of the number of letters
  
// algo:
//   - initialize an array which containing letters from ["a"..."z"]
//   - initialize an array count = [];
//   - iterate thru the array
//     - for each word, iterate thru the chars
//       - for each character, check its position: index + 1
//       - compare this position with the one in the letters array, 
//         - if equal, the count + 1
//         - otherwise, do nothing
//       - end loop
//     - end loop
//   - return the count array

const LETTERS = Array.from({ length: 26 }, (_, i) => String.fromCharCode('a'.charCodeAt(0) + i));

function symmetry_letters(array) {
  let result = [];
  
  array.map(word=> {
    let count = 0;
    
    word.split('').forEach((char, idx) => {
      let position = idx + 1;
      if (position === (LETTERS.indexOf(char.toLowerCase()) + 1)) count += 1;
    });
    
    result.push(count);
  });
  
  return result;
}

console.log(symmetry_letters(["abode","ABc","xyzD"])) // [4, 3, 1]
